File: change-finder/java/changefinder/Denomination.java
package changefinder;

public enum Denomination {
	
	fiftyPound(5000, "£50"),
	twentyPound(2000, "£20"),
	tenPound(1000, "£10"),
	fivePound(500, "£5"),
	twoPound(200, "£2"),
	onePound(100, "£1"),
	fiftyPence(50, "50p"),
	twentyPence(20, "20p"),
	tenPence(10, "10p"),
	fivePence(5, "5p"),
	twoPence(2, "2p"),
	onePence(1, "1p");

	private final int value;
	private final String display;

	Denomination(final int value, final String display) {
		this.value = value;
		this.display = display;
	}

	public int getValue() {
		return value;
	}

	@Override
	public String toString() {
		return display;
	}
}

File: change-finder/scala/Denomination.scala
package changefinder;

object Denomination {
	
	lazy val denominations = FiftyPound :: TwentyPound :: TenPound :: FiftyPound :: TwoPound :: OnePound ::
	                         FiftyPence :: TwentyPence :: TenPence :: FivePence :: TwoPence :: OnePence :: Nil
}

sealed trait Denomination {
  def value: Int	
}

abstract class BaseDenomination(val value: Int, val display: String) extends Denomination {
  override def toString = display
}
case object FiftyPound extends BaseDenomination(5000, "\u00A350")
case object TwentyPound extends BaseDenomination(2000, "\u00A320")
case object TenPound extends BaseDenomination(1000, "\u00A310")
case object FivePound extends BaseDenomination(500, "\u00A35")
case object TwoPound extends BaseDenomination(200, "\u00A32")
case object OnePound extends BaseDenomination(100, "\u00A31")
case object FiftyPence extends BaseDenomination(50, "50p")
case object TwentyPence extends BaseDenomination(20, "20p")
case object TenPence extends BaseDenomination(10, "10p")
case object FivePence extends BaseDenomination(5, "5p")
case object TwoPence extends BaseDenomination(2, "2p")
case object OnePence extends BaseDenomination(1, "1p")
